esp8266: handle rtc overflow

puuu opened the pull request:
This pull request should address #2097. Tested over 20 hours without problems.

Please reviewing. This fix requires, that `pyb_rtc_raw_us()` is called at least once within 7.45h. E.g. by calling `time.localtime()`, `machine.RTC().datetime()` or `machine_deepsleep()`. I believe, this is a typical use case, but maybe it is better to break out an additional function and hook it in some timer.
Any suggestions are welcome.

puuu committed:
esp8266/modpybrtc: Handle RTC overflow.

ESP-SDK system_get_rtc_time() returns uint32 and therefore overflow
about every 7:45h.  Let's write the last state of system_get_rtc_time()
in RTC mem and use it to check for overflow. This commit require running
pyb_rtc_get_us_since_2000() at least once within 7 hours to avoid
overflow.

dpgeorge commented:
I wonder if the RTC hardware keeps track of RTC counter overflows?  The system_get_rtc_time() function does only access a 32-bit RTC register but maybe there's more information available in the hardware?  If not then a patch like this would be welcome.  I guess it's enough to just write in the documentation that one must call time.localtime() often enough if the application is long-running.

puuu commented:
Yes, if the RTC hardware itself keeps track of RTC counter overflows, then this would be the best way to go. If anyone has more information/documentation about the RTC hardware and how to access it, please tell. Unfortunately I did not found any further information.

I had a sort look in the source code of “another popular ESP scripting language”. They check for overflow in a watchdog timer and count the overflows in a global variable...

pfalcon commented:
@puuu : We should not patch generic docs with port-specific details, any such details should go to esp8266/general.rst.

And the patch doesn't follow MicroPython codestyle https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md

pfalcon commented:
Also, I had a personal intuitive feeling that the code is more complicated than it could be, and reviewing existing code, I see that it both does the processing not in the best place, and can potentially lead to incorrect results. So, it's better to rewrite for all processing to happen in pyb_rtc_get_us_since_2000(). Feel free to leave that to us with @dpgeorge .

puuu committed:
docs/esp8266/general: Note about RTC overflow.

puuu commented:
@pfalcon, thanks for having a look through the PR. I rewrote  for all processing to happen in pyb_rtc_get_us_since_2000() and proper initialization.
IMHO, to save rtc memory, we have to check the raw RTC values. Doing this, there is no more need for `pyb_rtc_raw_us()`. That is why I removed the function.
At the moment I am doing a long-term test of this RP. Looks good so far.

@pfalcon and @dpgeorge, please have a look. And for sure feel free solve your way.

pfalcon commented:
This looks much better, thanks! I'd leave this for a more detailed review by @dpgeorge if he has a chance, or myself, but definitely would like to merge before Friday to fit into 1.8.1 release.

dpgeorge commented on esp8266/modpybrtc.c:
```diff
@@ -81,28 +84,40 @@ STATIC mp_obj_t pyb_rtc_make_new(const mp_obj_type_t *type, mp_uint_t n_args, mp
+    system_rtc_mem_write(MEM_OF_CHECK_ADDR, &rtc_ticks, sizeof(rtc_ticks));
+    if (rtc_ticks < rtc_last) {
+        // Adjust delta because of RTC overflow.
+        delta += ((uint64_t)1<<20) * cal;
```
This can be written simply as `(uint64_t)cal << 20`.

dpgeorge commented:
Yes it looks good.  @puuu please report if it works for you then we can merge.

puuu commented:
With the last modification I have it running for 12h+ without problems. 

@dpgeorge: I have realized that the RTC memory will survive resetting and flashing. Since we have changed the RTC memory structure, I wonder if we should change `MEM_MAGIC` to force the initializations? This may also necessary because of #2138. Does increasing of  `MEM_MAGIC` is OK, or will it have some side effects?

dpgeorge commented:
> With the last modification I have it running for 12h+ without problems.

Excellent, thanks for testing!

> I have realized that the RTC memory will survive resetting and flashing. Since we have changed the RTC memory structure, I wonder if we should change MEM_MAGIC to force the initializations?

RTC memory won't survive a power down so I think it's fine to keep MEM_MAGIC as-is.

dpgeorge commented:
There is a small issue with this as it stands: when you do a hard reset (pressing the reset button) the internal RTC clock (accessed via system_get_rtc_time()) is reset as well, even though the rest of the RTC state remains intact (ie all the state in the RTC memory including the magic number).  So if you get the time after a hard reset it always makes an adjustment by 7h:45min.

Maybe in mp_hal_rtc_init it should store the current system_get_rtc_time() in the MEM_OF_CHECK_ADDR slot regardless of whether the magic is set or not?

puuu commented:
@dpgeorge: Thank you for testing the reset case. Now I did some more test in this direction. It seems, the internal RTC clock always reset, except on soft reset.(CRTL-D, `machine.reset()`). Will say, also in case of deepsleep. This is really annoying, because the idea of using RTC memory was because of deepsleep. If the internal RTC clock will not survive deepsleep, why using RTC memory, even for `delta` and `cal`? IMHO ignoring `magic` will not help against a jumping clock after reset/deepsleep (either direction).

@pfalcon: IMHO, having 1.8.1 release in mind, this RP will solve #2097. Having a not working RTC time after reset/deepsleep is another, still remaining issue. I will have a deeper look into it.

dpgeorge commented:
Yes I agree that clock overflow while running is a distinct issue to clock restoration from deepsleep.  Merging this PR will fix the first problem.  Fixing the second issue is possible (eg you know how long the deepsleep should last so you can store in RTC memory the time that it will be when the device wakes, and then set the clock from that time).

So we can merge this as-is I"d say.

puuu commented:
I had a look here: http://bbs.espressif.com/viewtopic.php?t=1184#p4082
> ESP8266 designed like this, wake up from deep-sleep, the RTC timer will be reset, but the data stored in RTC memory will not.

Since there is no need for using RTC memory, I did a change to use a global variable. At the moment it is a separate commit, but I can rebase to clean up. @dpgeorge please can you have a look. Now I am testing it, need more 4 hour to be sure.

> Fixing the second issue is possible (eg you know how long the deepsleep should last so you can store in RTC memory the time that it will be when the device wakes, and then set the clock from that time).

I already wrote a patch, doing this by rewriting `delta`. I am planning to make a new pull request after testing it.

dpgeorge commented:
@puuu you can speed up the test dramatically by writing to the RTC counter itself:
```
import machine
machine.mem32[0x6000071c] = <new counter value>
```
Set the counter to something like 0xf0000000 to watch the overflow.

puuu commented:
> `machine.mem32[0x6000071c] = <new counter value>`

@dpgeorge: Thanks this sounds helpful, but unfortunately this address seems not to be writable, e.g. the value will not change.

Anyhow, now I can observe when the overflow occurs.

dpgeorge commented:
This patch looks good now, I think it can be (squashed) and merged.

dpgeorge commented on docs/esp8266/general.rst:
```diff
@@ -91,3 +91,12 @@ the following in ``main.py``::
+Due to limitations of the ESP8266 SDK, the internal RTC of the ESP8266
+overflow every 7.45h. If a longterm working RTC time is required, then
+``time()`` or ``localtime()`` must be called at least once within 7
+hours. Then MicroPython will handel the overflow.
```
This paragraph needs grammar/spelling fixes, but I can do it when merging.

puuu replied:
Please do so.  You may can also add a sentence that the RTC will/can not survive hard reset.

dpgeorge commented:
Great!  Merged in cafdfb7af3628aac5110770097b6cb4fdd70b9d5 and 3f6ca4b22ced848a4b849610a8bce0410c2091cc.
